### PR TITLE
Added utf8_to_ucs2 and utf8_to_ucs4

### PR DESCRIPTION
**Pull request by dmazzella**
Added utf8_to_ucs2 and utf8_to_ucs4 corresponding test vectors added

### CODE REVIEW

**Comment by dmazzella** [comment]
@randombit After this PR is merged, I create a new PR for pkcs12-kdf (charset is a dependency of pkcs12-kdf). Is this the correct workflow for you?

**Comment by randombit** [comment]
> After this PR is merged, I create a new PR for pkcs12-kdf (charset is a dependency of pkcs12-kdf). Is this the correct workflow for you?

Sure that works

File: src/lib/utils/charset.cpp [hunk 1]
```diff
@@ -67,6 +119,24 @@ std::string ucs2_to_utf8(const uint8_t ucs2[], size_t len) {
...
+
+   size_t pos = 0;
+   while(pos < utf8.size()) {
+      const uint32_t c = next_utf8_codepoint(utf8, pos);
+      if(c > 0xFFFF) {
+         throw Decoding_Error("Cannot encode character in UCS-2");
+      }
+      const uint16_t val = static_cast<uint16_t>(c);
+      out.push_back(static_cast<uint8_t>((val >> 8) & 0xFF));
+      out.push_back(static_cast<uint8_t>(val & 0xFF));
```
Comment (randombit) [inline src/lib/utils/charset.cpp]: Use `get_byte<0>` / `get_byte<1>` here

File: src/lib/utils/charset.cpp [hunk 2]
```diff
@@ -83,6 +153,22 @@ std::string ucs4_to_utf8(const uint8_t ucs4[], size_t len) {
...
+   std::vector<uint8_t> out;
+   out.reserve(utf8.size() * 4);
+
+   size_t pos = 0;
+   while(pos < utf8.size()) {
+      const uint32_t val = next_utf8_codepoint(utf8, pos);
+      out.push_back(static_cast<uint8_t>((val >> 24) & 0xFF));
+      out.push_back(static_cast<uint8_t>((val >> 16) & 0xFF));
+      out.push_back(static_cast<uint8_t>((val >> 8) & 0xFF));
+      out.push_back(static_cast<uint8_t>(val & 0xFF));
```
Comment (randombit) [inline src/lib/utils/charset.cpp]: `get_byte`

File: src/tests/data/charset.vec [hunk 3]
```diff
@@ -12,10 +12,96 @@ Out = EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB8AB2EB
 In = 0A0000000000000000000000000030000000000000001D1D1D1D01000000000000001D1D1D1D00000000000000000000
 Out = E0A880000000000000E38080000000E1B49DE1B49DC480000000E1B49DE1B49D0000000000
 
+[UTF8-UCS2]
```
Comment (randombit) [inline src/tests/data/charset.vec:15]: Thanks these tests looks good

Can you add also tests for correct handling of the minimum and maximum codepoints of a particular length?

Valid: C280, E0A080, F0908080, F48FBFBF, and also the empty string
Invalid: C279, ...

**Comment by randombit** [comment]
lgtm but please squash to 1 commit

**Comment by dmazzella** [comment]
> lgtm but please squash to 1 commit

done

**Review by randombit — APPROVED** [review]
